// @cyclonedx/cyclonedx-library/dist.node/enums/externalReferenceType.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalReferenceType = void 0;
var ExternalReferenceType;
(function (ExternalReferenceType) {
    ExternalReferenceType["VCS"] = "vcs";
    ExternalReferenceType["IssueTracker"] = "issue-tracker";
    ExternalReferenceType["Website"] = "website";
    ExternalReferenceType["Advisories"] = "advisories";
    ExternalReferenceType["BOM"] = "bom";
    ExternalReferenceType["MailingList"] = "mailing-list";
    ExternalReferenceType["Social"] = "social";
    ExternalReferenceType["Chat"] = "chat";
    ExternalReferenceType["Documentation"] = "documentation";
    ExternalReferenceType["Support"] = "support";
    ExternalReferenceType["Distribution"] = "distribution";
    ExternalReferenceType["License"] = "license";
    ExternalReferenceType["BuildMeta"] = "build-meta";
    ExternalReferenceType["BuildSystem"] = "build-system";
    ExternalReferenceType["ReleaseNotes"] = "release-notes";
    ExternalReferenceType["Other"] = "other";
})(ExternalReferenceType = exports.ExternalReferenceType || (exports.ExternalReferenceType = {}));
