// @cyclonedx/cyclonedx-library/dist.node/spec.js
"use strict";
/*!
This file is part of CycloneDX JavaScript Library.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Spec_version, _Spec_formats, _Spec_componentTypes, _Spec_hashAlgorithms, _Spec_hashValuePattern, _Spec_externalReferenceTypes, _Spec_supportsDependencyGraph, _Spec_supportsToolReferences, _Spec_requiresComponentVersion, _Spec_supportsProperties;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpecVersionDict = exports.Spec1dot4 = exports.Spec1dot3 = exports.Spec1dot2 = exports.UnsupportedFormatError = exports.Format = exports.Version = void 0;
const enums_1 = require("./enums");
var Version;
(function (Version) {
    Version["v1dot0"] = "1.0";
    Version["v1dot1"] = "1.1";
    Version["v1dot2"] = "1.2";
    Version["v1dot3"] = "1.3";
    Version["v1dot4"] = "1.4";
})(Version = exports.Version || (exports.Version = {}));
var Format;
(function (Format) {
    Format["XML"] = "xml";
    Format["JSON"] = "json";
})(Format = exports.Format || (exports.Format = {}));
class UnsupportedFormatError extends Error {
}
exports.UnsupportedFormatError = UnsupportedFormatError;
class Spec {
    constructor(version, formats, componentTypes, hashAlgorithms, hashValuePattern, externalReferenceTypes, supportsDependencyGraph, supportsToolReferences, requiresComponentVersion, supportsProperties) {
        _Spec_version.set(this, void 0);
        _Spec_formats.set(this, void 0);
        _Spec_componentTypes.set(this, void 0);
        _Spec_hashAlgorithms.set(this, void 0);
        _Spec_hashValuePattern.set(this, void 0);
        _Spec_externalReferenceTypes.set(this, void 0);
        _Spec_supportsDependencyGraph.set(this, void 0);
        _Spec_supportsToolReferences.set(this, void 0);
        _Spec_requiresComponentVersion.set(this, void 0);
        _Spec_supportsProperties.set(this, void 0);
        __classPrivateFieldSet(this, _Spec_version, version, "f");
        __classPrivateFieldSet(this, _Spec_formats, new Set(formats), "f");
        __classPrivateFieldSet(this, _Spec_componentTypes, new Set(componentTypes), "f");
        __classPrivateFieldSet(this, _Spec_hashAlgorithms, new Set(hashAlgorithms), "f");
        __classPrivateFieldSet(this, _Spec_hashValuePattern, hashValuePattern, "f");
        __classPrivateFieldSet(this, _Spec_externalReferenceTypes, new Set(externalReferenceTypes), "f");
        __classPrivateFieldSet(this, _Spec_supportsDependencyGraph, supportsDependencyGraph, "f");
        __classPrivateFieldSet(this, _Spec_supportsToolReferences, supportsToolReferences, "f");
        __classPrivateFieldSet(this, _Spec_requiresComponentVersion, requiresComponentVersion, "f");
        __classPrivateFieldSet(this, _Spec_supportsProperties, supportsProperties, "f");
    }
    get version() {
        return __classPrivateFieldGet(this, _Spec_version, "f");
    }
    supportsFormat(f) {
        return __classPrivateFieldGet(this, _Spec_formats, "f").has(f);
    }
    supportsComponentType(ct) {
        return __classPrivateFieldGet(this, _Spec_componentTypes, "f").has(ct);
    }
    supportsHashAlgorithm(ha) {
        return __classPrivateFieldGet(this, _Spec_hashAlgorithms, "f").has(ha);
    }
    supportsHashValue(hv) {
        return typeof hv === 'string' &&
            __classPrivateFieldGet(this, _Spec_hashValuePattern, "f").test(hv);
    }
    supportsExternalReferenceType(ert) {
        return __classPrivateFieldGet(this, _Spec_externalReferenceTypes, "f").has(ert);
    }
    get supportsDependencyGraph() {
        return __classPrivateFieldGet(this, _Spec_supportsDependencyGraph, "f");
    }
    get supportsToolReferences() {
        return __classPrivateFieldGet(this, _Spec_supportsToolReferences, "f");
    }
    get requiresComponentVersion() {
        return __classPrivateFieldGet(this, _Spec_requiresComponentVersion, "f");
    }
    supportsProperties() {
        return __classPrivateFieldGet(this, _Spec_supportsProperties, "f");
    }
}
_Spec_version = new WeakMap(), _Spec_formats = new WeakMap(), _Spec_componentTypes = new WeakMap(), _Spec_hashAlgorithms = new WeakMap(), _Spec_hashValuePattern = new WeakMap(), _Spec_externalReferenceTypes = new WeakMap(), _Spec_supportsDependencyGraph = new WeakMap(), _Spec_supportsToolReferences = new WeakMap(), _Spec_requiresComponentVersion = new WeakMap(), _Spec_supportsProperties = new WeakMap();
exports.Spec1dot2 = Object.freeze(new Spec(Version.v1dot2, [
    Format.XML,
    Format.JSON
], [
    enums_1.ComponentType.Application,
    enums_1.ComponentType.Framework,
    enums_1.ComponentType.Library,
    enums_1.ComponentType.Container,
    enums_1.ComponentType.OperatingSystem,
    enums_1.ComponentType.Device,
    enums_1.ComponentType.Firmware,
    enums_1.ComponentType.File
], [
    enums_1.HashAlgorithm.MD5,
    enums_1.HashAlgorithm['SHA-1'],
    enums_1.HashAlgorithm['SHA-256'],
    enums_1.HashAlgorithm['SHA-384'],
    enums_1.HashAlgorithm['SHA-512'],
    enums_1.HashAlgorithm['SHA3-256'],
    enums_1.HashAlgorithm['SHA3-384'],
    enums_1.HashAlgorithm['SHA3-512'],
    enums_1.HashAlgorithm['BLAKE2b-256'],
    enums_1.HashAlgorithm['BLAKE2b-384'],
    enums_1.HashAlgorithm['BLAKE2b-512'],
    enums_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    enums_1.ExternalReferenceType.VCS,
    enums_1.ExternalReferenceType.IssueTracker,
    enums_1.ExternalReferenceType.Website,
    enums_1.ExternalReferenceType.Advisories,
    enums_1.ExternalReferenceType.BOM,
    enums_1.ExternalReferenceType.MailingList,
    enums_1.ExternalReferenceType.Social,
    enums_1.ExternalReferenceType.Chat,
    enums_1.ExternalReferenceType.Documentation,
    enums_1.ExternalReferenceType.Support,
    enums_1.ExternalReferenceType.Distribution,
    enums_1.ExternalReferenceType.License,
    enums_1.ExternalReferenceType.BuildMeta,
    enums_1.ExternalReferenceType.BuildSystem,
    enums_1.ExternalReferenceType.Other
], true, false, true, false));
exports.Spec1dot3 = Object.freeze(new Spec(Version.v1dot3, [
    Format.XML,
    Format.JSON
], [
    enums_1.ComponentType.Application,
    enums_1.ComponentType.Framework,
    enums_1.ComponentType.Library,
    enums_1.ComponentType.Container,
    enums_1.ComponentType.OperatingSystem,
    enums_1.ComponentType.Device,
    enums_1.ComponentType.Firmware,
    enums_1.ComponentType.File
], [
    enums_1.HashAlgorithm.MD5,
    enums_1.HashAlgorithm['SHA-1'],
    enums_1.HashAlgorithm['SHA-256'],
    enums_1.HashAlgorithm['SHA-384'],
    enums_1.HashAlgorithm['SHA-512'],
    enums_1.HashAlgorithm['SHA3-256'],
    enums_1.HashAlgorithm['SHA3-384'],
    enums_1.HashAlgorithm['SHA3-512'],
    enums_1.HashAlgorithm['BLAKE2b-256'],
    enums_1.HashAlgorithm['BLAKE2b-384'],
    enums_1.HashAlgorithm['BLAKE2b-512'],
    enums_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    enums_1.ExternalReferenceType.VCS,
    enums_1.ExternalReferenceType.IssueTracker,
    enums_1.ExternalReferenceType.Website,
    enums_1.ExternalReferenceType.Advisories,
    enums_1.ExternalReferenceType.BOM,
    enums_1.ExternalReferenceType.MailingList,
    enums_1.ExternalReferenceType.Social,
    enums_1.ExternalReferenceType.Chat,
    enums_1.ExternalReferenceType.Documentation,
    enums_1.ExternalReferenceType.Support,
    enums_1.ExternalReferenceType.Distribution,
    enums_1.ExternalReferenceType.License,
    enums_1.ExternalReferenceType.BuildMeta,
    enums_1.ExternalReferenceType.BuildSystem,
    enums_1.ExternalReferenceType.Other
], true, false, true, true));
exports.Spec1dot4 = Object.freeze(new Spec(Version.v1dot4, [
    Format.XML,
    Format.JSON
], [
    enums_1.ComponentType.Application,
    enums_1.ComponentType.Framework,
    enums_1.ComponentType.Library,
    enums_1.ComponentType.Container,
    enums_1.ComponentType.OperatingSystem,
    enums_1.ComponentType.Device,
    enums_1.ComponentType.Firmware,
    enums_1.ComponentType.File
], [
    enums_1.HashAlgorithm.MD5,
    enums_1.HashAlgorithm['SHA-1'],
    enums_1.HashAlgorithm['SHA-256'],
    enums_1.HashAlgorithm['SHA-384'],
    enums_1.HashAlgorithm['SHA-512'],
    enums_1.HashAlgorithm['SHA3-256'],
    enums_1.HashAlgorithm['SHA3-384'],
    enums_1.HashAlgorithm['SHA3-512'],
    enums_1.HashAlgorithm['BLAKE2b-256'],
    enums_1.HashAlgorithm['BLAKE2b-384'],
    enums_1.HashAlgorithm['BLAKE2b-512'],
    enums_1.HashAlgorithm.BLAKE3
], /^([a-fA-F0-9]{32})$|^([a-fA-F0-9]{40})$|^([a-fA-F0-9]{64})$|^([a-fA-F0-9]{96})$|^([a-fA-F0-9]{128})$/, [
    enums_1.ExternalReferenceType.VCS,
    enums_1.ExternalReferenceType.IssueTracker,
    enums_1.ExternalReferenceType.Website,
    enums_1.ExternalReferenceType.Advisories,
    enums_1.ExternalReferenceType.BOM,
    enums_1.ExternalReferenceType.MailingList,
    enums_1.ExternalReferenceType.Social,
    enums_1.ExternalReferenceType.Chat,
    enums_1.ExternalReferenceType.Documentation,
    enums_1.ExternalReferenceType.Support,
    enums_1.ExternalReferenceType.Distribution,
    enums_1.ExternalReferenceType.License,
    enums_1.ExternalReferenceType.BuildMeta,
    enums_1.ExternalReferenceType.BuildSystem,
    enums_1.ExternalReferenceType.ReleaseNotes,
    enums_1.ExternalReferenceType.Other
], true, true, false, true));
exports.SpecVersionDict = Object.freeze({
    [Version.v1dot2]: exports.Spec1dot2,
    [Version.v1dot3]: exports.Spec1dot3,
    [Version.v1dot4]: exports.Spec1dot4
});
